pkg/certrotation/target.go

```diff
@@ -0,0 +1,141 @@
+
+	"github.com/openshift/library-go/pkg/crypto"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/cert"
+)
+
+// TargetRotation rotates a key and cert signed by a CA. It creates a new one when 80%
+// of the lifetime of the old cert has passed, or the CA used to signed the old cert is
+// gone from the CA bundle.
+type TargetRotation struct {
+	Namespace     string
+	Name          string
+	Validity      time.Duration
+	HostNames     []string
+	Lister        corev1listers.SecretLister
+	Client        corev1client.SecretsGetter
+	EventRecorder events.Recorder
+}
+
+func (c TargetRotation) EnsureTargetCertKeyPair(signingCertKeyPair *crypto.CA, caBundleCerts []*x509.Certificate) error {
+	originalTargetCertKeyPairSecret, err := c.Lister.Secrets(c.Namespace).Get(c.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	targetCertKeyPairSecret := originalTargetCertKeyPairSecret.DeepCopy()
+	if apierrors.IsNotFound(err) {
+		// create an empty one
+		targetCertKeyPairSecret = &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: c.Namespace, Name: c.Name}}
+	}
+	targetCertKeyPairSecret.Type = corev1.SecretTypeTLS
+
+	reason := needNewTargetCertKeyPair(targetCertKeyPairSecret, caBundleCerts)
+	if len(reason) == 0 {
+		return nil
+	}
+
+	c.EventRecorder.Eventf("TargetUpdateRequired", "%q in %q requires a new target cert/key pair: %v", c.Name, c.Namespace, reason)
+	if err := c.setTargetCertKeyPairSecret(targetCertKeyPairSecret, c.Validity, signingCertKeyPair); err != nil {
+		return err
+	}
+
+	if _, _, err = resourceapply.ApplySecret(context.TODO(), c.Client, c.EventRecorder, targetCertKeyPairSecret); err != nil {
+		return err
+	}
+	return nil
+}
+
+// needNewTargetCertKeyPair returns a reason for creating a new target cert/key pair.
+// Return empty if a valid cert/key pair is in place and no need to rotate it yet.
+//
+// We create a new target cert/key pair if
+//   1) no cert/key pair exits
+//   2) or the cert expired (then we are also pretty late)
+//   3) or we are over the renewal percentage of the validity
+//   4) or our old CA is gone from the bundle (then we are pretty late to the renewal party)
+func needNewTargetCertKeyPair(secret *corev1.Secret, caBundleCerts []*x509.Certificate) string {
+	certData := secret.Data["tls.crt"]
+	if len(certData) == 0 {
+		return "missing tls.crt"
+	}
+
+	certificates, err := cert.ParseCertsPEM(certData)
+	if err != nil {
+		return "bad certificate"
+	}
+	if len(certificates) == 0 {
+		return "missing certificate"
+	}
+
+	cert := certificates[0]
+	if time.Now().After(cert.NotAfter) {
+		return "already expired"
+	}
+
+	maxWait := cert.NotAfter.Sub(cert.NotBefore) / 5
+	latestTime := cert.NotAfter.Add(-maxWait)
+	if time.Now().After(latestTime) {
```

Comment:
https://github.com/open-cluster-management-io/registration/blob/ef7c23ededb2f886fc5e799d2125637709ba1688/pkg/clientcert/cert_controller.go#L199

it should be jittered so that certs can be rotated progressively, otherwise all the old certs will be wiped out at the same moment (which is practically dangerous)?

Reply:
Here is a example of usage:
```
for _, targetRotation := range c.targetRotations {
		if err := targetRotation.EnsureTargetCertKeyPair(signingCertKeyPair, cabundleCerts); err != nil {
			errs = append(errs, err)
		}
	}
```
We can see that `EnsureTargetCertKeyPair` is invoked one by one, so as `needNewTargetCertKeyPair` and `setTargetCertKeyPairSecret`. 
So actually it won't update all old certs at the same time.